### ISSUE TITLE
Connection Suspension

### ISSUE THREAD

**Issue by Stebalien** [libp2p/go-libp2p#438]
Currently, we keep a bunch of connections open because re-opening connections is expensive. However this costs us in terms of memory and background CPU usage.

It would be nice if we could "suspend" some connections. That is, close all streams and *act* like we're disconnected but leave the underlying connection open.

This will likely require an actual connection suspension protocol so both sides can agree that the connection should be "put on ice". Taking it *off* ice would be a simple matter of using it.

Note: this won't fix the issue of file descriptor use. For that, we'll still need a UDP based protocol like QUIC.

---

Note: Ideally, we'd keep fewer background streams open *anyways*. If we improved this situation, we may not need connection suspension. However, it's still useful to have a concept of connections that should only be used "if need be".

**Comment by raulk** [comment]
We could have SUSPENDED and HIBERNATING states. 

Suspended would be as you described, but upon hibernation, both parties could exchange a session token they can use to bypass SecIO and protocol negotiation when they reconnect. The token would have a TTL agreed by both parties during the HIBERNATE message exchange. We could store the token in the peerstore.

**Comment by Stebalien** [comment]
> The token would have a TTL agreed by both parties during the HIBERNATE message exchange (e.g. the lesser of two proposals).

IMO, *this* should be handled at the transport layer (otherwise, we need to do something special for each transport or come up with some general-purpose interface).

Additionally, we probably don't need to make it explicit. Really, we can transports can just do it automatically on connect. Both QUIC and TLS-1.3 have session resumption and we can probably just have them stash session-resumption information in the peerstore without any additional help.

Basically, I agree but we can probably make this *mostly* transparent (I think?).

**Comment by h1z1** [comment]
There's no inherent requirement in TCP or UDP to immiediately respond.  There is however a bug in GO/IPFS where connections close prematurely.  Fixing that would allow the OS to properly queue data.

**Comment by raulk** [comment]
> IMO, this should be handled at the transport layer

@Stebalien – you're right. Perhaps we can add a `Resumable()` method to `Transport`, so that transports can declare whether they can resume connections natively. Then the `Upgrader` could add a shim in the form of an explicit protocol if the underlying transport is not resumable.

Actually, rather than `Transport.Resumable()` I'd make it future-proof by adding a `Transport.Features()` that returns a `map[int]int` (or enum-ish type alias instead of the key `int`) so consumers can do:

```
var t tpt.Transport
c := t.Features()
if c[tpt.FeatResumable] == FeatValueTrue {
   ...
}
```

Or maybe `IsSupported(tpt.FeatResumable)` to avoid map copies.

**Comment by raulk** [comment]
> There's no inherent requirement in TCP or UDP to immiediately respond. There is however a bug in GO/IPFS where connections close prematurely. Fixing that would allow the OS to properly queue data.

@h1z1 - You're probably referring to the connection manager, right? Keeping the connection count in check is important to avoid abusing resources. Suspending connections would bring down the cost of re-establishing connections closed by the connection manager.

In parallel, we're also aiming for a more intelligent connection manager. See https://github.com/libp2p/go-libp2p-connmgr/issues.

**Comment by h1z1** [comment]
> 
> @h1z1 - You're probably referring to the connection manager, right? Keeping the connection count in check is important to avoid abusing resources.
> 
> Suspending connections would bring down the cost of re-establishing connections closed by the connection manager.
> 

If I understand above, suspend in this case would be the same as if you interrupted a connection (either intentional via firewall rules or otherwise) no?  With the addition of an explicit state of course.

> In parallel, we're also aiming for a more intelligent connection manager. See https://github.com/libp2p/go-libp2p-connmgr/issues.

**Comment by Stebalien** [comment]
> If I understand above, suspend in this case would be the same as if you interrupted a connection (either intentional via firewall rules or otherwise) no? With the addition of an explicit state of course.

Not quite. We'd keep the connection open, we just wouldn't use it (except for occasional keep-alives). Well, really, how we do this will be up to the transport.

> Actually, rather than Transport.Resumable() I'd make it future-proof by adding a Transport.Features() that returns a map[int]int (or enum-ish type alias instead of the key int) so consumers can do:

We can implement most features with interfaces. That is, we can add a `Suspend` method to `go-libp2p-net.Conn` and have it *try* casting the internal connection to a `SuspendableConn` interface. However, in practice, we may not even have to do that. All we *really* need to do is kill all the open streams (at least for TCP connections). That will effectively "suspend" the connection. The nice thing about doing this at the swarm level is that it's transport agnostic.

What we *might* want is some form of "fast-reconnect" capability. That is, if the swarm determines that the connection can be quickly re-established, it can just kill it completely.  For example, re-establishing a QUIC connection can be free with 0-RTT and session tokens (which we can store in the peerstore). Personally, I'd add a `Stat` method to the `go-libp2p-transport.Conn` (like the one on `go-libp2p-net.Conn`) to handle this but we could also use a `Supports(feature)` method.

**Comment by Stebalien** [comment]
> There are a couple issues with QUIC mostly due to the lack of local administrative control. It works in Google for the same reason bittorrent does/did, non-commercial firewalls don't know what to do with the traffic behond being a generic udp flow. Also suffers the same problems mainly congestion and lack of localized retransmits (a lost packet is a full rtt). There's no way to bind different flows to different paths unlike tcp..etc.

This is actually exactly why we like QUIC. The hope is that crappy consumer routers won't try doing fancy things like connection tracking etc. and just forward the packets. One of our *biggest* problems with keeping so many connections open is that consumer routers (a) try to keep track of every connection and then (b) crash or freeze because we open too many.

**Comment by Stebalien** [comment]
> Missed the second part of that, it's not just those routers though. Any load balanced link upstream from them could be (ok very likely) using some form of hashing to distribute the load. Typically that is based on port and ip. It becomes more of an issue at higher speeds.

They can still do that with UDP (they can't track individual connections but they can track "flows") and I'm less worried about upstream hardware; it's just consumer-grade hardware that tends to be really crappy.

> There's a lot of garbage that comes with running a node, no way I'd ever do it without a stateful firewall in front.

We're working on reducing background traffic but background DHT traffic will always be pretty high. FYI, you can put your DHT node into client mode (using the `Client` option) if you just want to use the DHT but don't want to answer DHT queries.

**Comment by h1z1** [comment]
> 
> They can still do that with UDP (they can't track individual connections but they can track "flows") and I'm less worried about upstream hardware; it's just consumer-grade hardware that tends to be really crappy.
> 

Not sure I follow, the multiplex happens above udp, hashing is typically based on lower variables (port / src/dst host, etc).  The kernel only sees one flow.   The same thing happens in consumer hardware with things like RPS.  One of the interesting side effects not really addressed by Google is contention in the demux.

> > There's a lot of garbage that comes with running a node, no way I'd ever do it without a stateful firewall in front.
> 
> We're working on reducing background traffic but background DHT traffic will always be pretty high. FYI, you can put your DHT node into client mode (using the `Client` option) if you just want to use the DHT but don't want to answer DHT queries.

Indeed, I've done that :)   DHT nodes are running on bastion hosts.

**Comment by Stebalien** [comment]
This is getting way off topic, let's open a new issue to discuss QUIC and the implications (here: https://github.com/libp2p/libp2p). Note: We're not going to deprecate or even disable the TCP transport so you'll still be able to use it.

**Comment by jjjachyty** [comment]
@Stebalien running nodeA and nodeB、nodeC,nodeB connect to NodeA,NodeC connect nodeA，when i shutdown NodeB，NodeA immediately close，and NodeC also immediately close,how todo shutdown NodeB and NodeC->NodeA was keep connnect？
